### DISCUSSION TITLE
Getting SHAP to work with an unusual embedding input

### DISCUSSION THREAD

**Discussion by jlapin1** [shap/shap#2563] [Q&A]
Hello SHAP community

I would really like to get the library working to explain with the input embedding matrix for my model. Our input is a 40x40 matrix of several stacked one-hot vectors and a float. Specifically these are protein sequences with a max length of 40 amino acids. Each amino acid is represented by a 21 length one-hot vector, each amino acid can have 1 of 11 types of chemical modifications (length 11 one-hot vector, the entire protein has 1 of 8 charges (length 8 one-hot for all 40 columns/amino acids), and finally a float for an energy (replicated across all 40 columns). Just to restate this, each column of this embedding matrix is stacked amino acid/chemical modification/charge/energy.

I could represent each input as a length 42 index vector, where the first 40 token indices represent a combination of amino acid and modification type, and the last 2 vector indices are the charge and energy. This could then be converted to the 40x40 input embedding and run through the network, which outputs a length 20k vector to be compared to a target vector with cosine similarity, resulting in a score from 0-1. 

For the experts with implementing SHAP on deep models, how would you approach this problem? Ultimately I believe there is some way to get this to work so that each explanation outputs 42 shapley values to explain the cosine similarity.

Thank you in advance
Joel

**Comment by jainraj** [comment]
You could probably use Gradient Explainer - this would give SHAP value for each input dimension and then hopefully you can use these values.

**Comment by jlapin1** [comment]
Thank you Raj. I actually solved this problem months ago by creating an indexed token for each input. Specifically, I designated a token for each combination of 20 amino acids and 8 modifications, which can then be turned into the corresponding one-hot vectors, and ultimately a machine readable input tensor to the model. This way the SHAP library could sample amongst these tokens to produce an estimate for the SHAP values for each of the 40 (max sequence length) inputs.

This required a lot of coding to make it compatible with the SHAP libraries' functions. I needed to write a subroutine to batch the inputs since the GPU can only handle so much data. I also needed to write a subroutine to handle all the aforementioned token-to-input-tensor work.

In general the SHAP values were consistent with our understanding of the underlying scientific phenomena, which was great, but the accuracy was limited by the amount of sampling that I could feasibly do in a practical time period. All those forward passes of the model really do slow the program down.
